packages/grid/x-data-grid/src/internals/hooks/features/keyboard/useGridKeyboardNavigation.ts

```diff
@@ -167,12 +170,31 @@ export const useGridKeyboardNavigation = (
   );
 
   const handleColumnHeaderKeyDown = React.useCallback<
-    GridEventListener<GridEvents.columnHeaderNavigationKeyDown>
+    GridEventListener<GridEvents.columnHeaderKeyDown>
   >(
     (params, event) => {
-      if (!params.field) {
+      let isFromCustomHeader = false;
+      if (params.colDef.renderHeader) {
+        const headerTitleNode = event.currentTarget.querySelector(
+          `.${gridClasses.columnHeaderTitleContentContainer}`,
+        );
+        isFromCustomHeader =
+          !!headerTitleNode && headerTitleNode.contains(event.target as Node | null);
+      }
+
+      if (
+        !isGridHeaderCellRoot(event.target as HTMLElement) &&
```

Comment:
This helper is not used elsewhere. We want to remove it in v6. Could you inline it here?